framework/include/utils/RankTwoTensor.h

```diff
@@ -262,149 +550,703 @@ class RankTwoTensorTempl : public TensorValue<T>
+   * RankTwoTensor A(1, 2, 3, 4, 5, 6, 7, 8, 9);
+   * // A = [ 1 2 3
+   * //       2 4 6
+   * //       3 6 9 ]
+   * RealVectorValue a(1, 2, 3);
+   * // a = [ 1
+   * //       2
+   * //       3 ]
+   * RealVectorValue b = A * a;
+   * // b = [ 30
+   * //       36
+   * //       42 ]
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   */
   template <typename T2>
-  TypeVector<typename CompareTypes<T, T2>::supertype> operator*(const TypeVector<T2> & a) const;
+  libMesh::TypeVector<typename CompareTypes<T, T2>::supertype>
+  operator*(const libMesh::TypeVector<T2> & a) const;
 
-  /// Defines multiplication with a TypeTensor<T>
+  /**
+   * Multiplication with another second order tensor
+   *
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.cpp}
+   * RankTwoTensor A(1, 2, 3, 4, 5, 6, 7, 8, 9);
+   * // A = [ 1 2 3
+   * //       2 4 6
+   * //       3 6 9 ]
+   * RankTwoTensor B(9, 8, 7, 6, 5, 4, 3, 2, 1);
+   * // B = [ 9 6 3
+   * //       8 5 2
+   * //       7 4 1 ]
+   * RankTwoTensor C = A * B;
+   * // C = [ 90  54 18
+   * //       114 69 24
+   * //       138 84 30 ]
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   */
   template <typename T2>
   RankTwoTensorTempl<typename CompareTypes<T, T2>::supertype>
-  operator*(const TypeTensor<T2> & a) const;
-
-  /// Defines multiplication with a TypeTensor<T>
-  RankTwoTensorTempl<T> & operator*=(const TypeTensor<T> & a);
+  operator*(const libMesh::TypeTensor<T2> & a) const;
 
-  /// Defines logical equality with another RankTwoTensorTempl<T>
-  bool operator==(const RankTwoTensorTempl<T> & a) const;
-
-  /// Sets _coords to the values in a ColumnMajorMatrix (must be 3x3)
-  RankTwoTensorTempl<T> & operator=(const ColumnMajorMatrixTempl<T> & a);
-
-  /// returns _coords_ij * a_ij (sum on i, j)
+  /**
+   * Return the double contraction with another second order tensor \f$ A_{ij} B_{ij} \f$
+   *
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.cpp}
+   * RankTwoTensor A(1, 2, 3, 4, 5, 6, 7, 8, 9);
+   * // A = [ 1 2 3
+   * //       2 4 6
+   * //       3 6 9 ]
+   * RankTwoTensor B(9, 8, 7, 6, 5, 4, 3, 2, 1);
+   * // B = [ 9 6 3
+   * //       8 5 2
+   * //       7 4 1 ]
+   * Real result = A.doubleContraction(B);
+   * // result = 143
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   */
   T doubleContraction(const RankTwoTensorTempl<T> & a) const;
 
   /**
-   * This function returns
-   * C_{ijkl} = a_{M(n)M(o)} b_{M(p)M(q)}.
-   * where the set of indices {n,o,p,q} is a permutation of the indices {i,j,k,l}
+   * Return the general tensor product of this second order tensor and another second order tensor
+   * defined as \f$ C_{ijkl} = A_{\mathcal{M}(n)\mathcal{M}(o)} B_{\mathcal{M}(p)\mathcal{M}(q)} \f$
+   * where the multiplication order is defined by the index map \f$ \mathcal{M}: \{n,o,p,q\} \to
+   * \{i,j,k,l\} \f$. The index map is specified using the template parameters. See examples below
```

Comment:
This drops the requirement that the mnop indices need to be a permutation ijkl (formerly described as a bijective map). This should appear in the docs in some form. If the requirement is not fulfilled  the result might be unexpected to the user.